tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -11716,200 +11253,153 @@ void ExecutionTest::LongVectorOpTestBase(
 
-// RunShaderOpTest is a helper function that handles resource creation
-// and setup. It also handles the shader compilation and execution. It takes a
-// callback that is called when the shader is compiled, but before it is
-// executed.
-std::shared_ptr<ShaderOpTestResult> TestResult = RunShaderOpTest(
-    D3DDevice, m_support, TestXML, "LongVectorOp",
-    [&](LPCSTR Name, std::vector<BYTE> &ShaderData, st::ShaderOp *ShaderOp) {
-      LogCommentFmt(L"RunShaderOpTest CallBack. Resource Name: %S", Name);
+  if (TestConfig.IsScalarOp())
+    // Scalar ops are always binary ops. So InputVector2ValueSet is initialized
+    // with values above.
+    ScalarInput[0] = InputVector2ValueSet[0];
 
-      // This callback is called once for each resource defined for
-      // "LongVectorOp" in ShaderOpArith.xml. All callbacks are fired for each
-      // resource. We determine whether they are applicable to the test case
-      // when they run.
+  // Fill the input vectors with values from the value set. Repeat the values
+  // when we reach the end of the value set.
+  for (size_t Index = 0; Index < N; Index++) {
+    InputVector1[Index] =
+        InputVector1ValueSet[Index % InputVector1ValueSet.size()];
 
-      // Process the callback for the OutputVector resource.
-      if (0 == _stricmp(Name, "OutputVector")) {
-        // We only need to set the compiler options string once. So this is a
-        // convenient place to do it.
-        ShaderOp->Shaders.at(0).Arguments = CompilerOptionsString.c_str();
+    if (IsVectorBinaryOp)
+      InputVector2[Index] =
+          InputVector2ValueSet[Index % InputVector2ValueSet.size()];
+  }
 
-        return;
-      }
+  if (TestConfig.HasInputArguments()) {
+    InputArgsArray = TestConfig.GetInputArgsArray();
+  }
 
-      // Process the callback for the InputFuncArgs resource.
-      if (0 == _stricmp(Name, "InputFuncArgs")) {
-        if (TestConfig.IsScalarOp) {
-          FillShaderBufferFromLongVectorData<T, 1>(ShaderData, ScalarInput);
-        } else if (TestConfig.OpType == LongVectorOpType_Clamp) {
-          std::array<T, 2> ClampArgs = {ClampArgMin, ClampArgMax};
-          FillShaderBufferFromLongVectorData<T, 2>(ShaderData, ClampArgs);
+  std::array<T, N> ExpectedVector;
+  if (IsVectorBinaryOp)
+    ExpectedVector =
+        ComputeExpectedValues(InputVector1, InputVector2, TestConfig);
+  else if (TestConfig.IsScalarOp())
+    ExpectedVector =
+        ComputeExpectedValues(InputVector1, ScalarInput[0], TestConfig);
+  else // Must be a unary op
+    ExpectedVector = ComputeExpectedValues(InputVector1, TestConfig);
+
+  if (LogInputs) {
+    LogLongVector<T, N>(InputVector1, L"InputVector1");
+
+    if (IsVectorBinaryOp)
+      LogLongVector<T, N>(InputVector2, L"InputVector2");
+    else if (TestConfig.IsScalarOp())
+      LogLongVector<T, 1>(ScalarInput, L"ScalarInput");
+
+    if (TestConfig.GetUnaryOpType() == LongVectorUnaryOpType_Clamp) {
```

Comment:
This should be more generic. I'll update it when there are more test cases with input args.